akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala

```diff
@@ -0,0 +1,106 @@
+ * It offers additional hooks to modify the default behavior
+ */
+trait HttpApp {
+
+  def main(args: Array[String]) = {
+    implicit val system = createActorSystem()
+    implicit val materializer = ActorMaterializer()
+    implicit val executionContext = system.dispatcher
+    postStartActorSystem(system)
+
+    val config = retrieveConfig()
+    val host = config.getString("akka.http.server.host")
+    val port = config.getInt("akka.http.server.port")
+
+    val bindingFuture = Http().bindAndHandle(
+      handler = route,
+      interface = host,
+      port = port,
+      settings = ServerSettings(config))
+
+    println(s"Server online at http://$host:$port/")
+
+    //It's waiting forever because maybe there is never a shutdown signal
+    Await.result(waitForShutdownSignal(system), Duration.Inf)
+
+    bindingFuture
+      .flatMap(_.unbind())
+      .onComplete(_ ⇒ {
+        preTerminateActorSystem(system)
+        system.terminate()
+      })
+  }
+
+  /**
+   * Provides the [[ActorSystem]] from where the routes will be served. Override this method, if the default [[ActorSystem]] doesn't suit your needs. You don't need to store the created instance of the [[ActorSystem]].
+   */
+  def createActorSystem(): ActorSystem = ActorSystem("my-system")
+
+  /**
+   * Hook that will be called once the actor system is created. Override this method if you want to do something with this [[ActorSystem]] like creating actors.
+   */
+  def postStartActorSystem(actorSystem: ActorSystem): Unit = {
+  }
+
+  /**
+   * Hook that will be called just before the termination of the actor system. Override this methods if you want to perform some actions on the current actorSystem, like gracefully shutting down any created actor.
+   */
+  def preTerminateActorSystem(actorSystem: ActorSystem): Unit = {
+  }
+
+  /**
+   * Hook that lets the user specify the future that will signal the shutdown of the server whenever completed
+   */
+  def waitForShutdownSignal(actorSystem: ActorSystem)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    Future {
+      println("Press RETURN to stop...")
+      StdIn.readLine()
+    }
+  }
+
+  /**
+   * Your class should provide an implementation for the routes. These will be bound to the [[Http]]
+   */
+  val route: Route
+
+  /**
+   * It loads the config provided by `configProvider` and add the needed defaults just in case they are not present
+   */
+  private def retrieveConfig(): Config = {
+    val config = createConfig()
+    val defaults = ConfigFactory.parseString(
+      """
+        |akka.http.server{
+        | host = localhost
+        | port = 8080
+        |}
+      """.stripMargin)
+    config.withFallback(defaults)
+  }
+
+  /**
+   * Enables the user the possibility to override the `Config` that should be used for the application. By default it loads the config in the path
+   */
+  def createConfig(): Config = {
+    ConfigFactory.load()
```

Comment:
Oops! Here's a severe bug! Well, it's up there where the actor system is created ... I'll add a comment up there.